File: Classes/IMQuickSearch.m
//
//  IMQuickSearch.m
//  IMQuickSearch
//
//  Created by Ben Gordon on 12/13/13.
//  Copyright (c) 2013 Intermark. All rights reserved.
//

#import "IMQuickSearch.h"

@implementation IMQuickSearch

#pragma mark - Init
- (instancetype)initWithFilters:(NSArray *)filters {
    // Fuzziness is not implemented due to speed considerations
    // This value does not matter at all.
    return [self initWithFilters:filters fuzziness:0.0];
}

- (instancetype)initWithFilters:(NSArray *)filters fuzziness:(float)fuzziness {
    if (self = [super init]) {
        self.masterArray = filters;
        self.fuzziness = fuzziness;
    }
    
    return self;
}


#pragma mark - Filter
- (NSArray *)filteredObjectsWithValue:(id)value {
    // Set Up Filter
    NSMutableSet *filteredSet = [NSMutableSet set];
    
    // Create copy of array to prevent mutability of filters mid-search
    NSArray *copyMasterArray = [self.masterArray copy];
    
    // Filter each object based on value
    for (IMQuickSearchFilter *quickFilter in copyMasterArray) {
        [filteredSet addObjectsFromArray:[quickFilter filteredObjectsWithValue:value]];
    }
    
    // Return array from set
    return [filteredSet allObjects];
}

- (void)asynchronouslyFilterObjectsWithValue:(id)value completion:(void (^)(NSArray *filteredResults))completion {
    // Start another thread
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSArray *filteredObjects = [self filteredObjectsWithValue:value];
        // Get Main Thread
        dispatch_async(dispatch_get_main_queue(), ^{
            completion(filteredObjects);
        });
    });
}


#pragma mark - Filter Management
- (void)addFilter:(IMQuickSearchFilter *)filter {
    if (filter) {
        NSMutableArray *newMasterArray = [self.masterArray mutableCopy];
        [newMasterArray addObject:filter];
        self.masterArray = newMasterArray;
    }
}

- (void)removeFilter:(IMQuickSearchFilter *)filter {
    if (filter) {
        NSMutableArray *newMasterArray = [self.masterArray mutableCopy];
        [newMasterArray removeObject:filter];
        self.masterArray = newMasterArray;
    }
}

@end

File: Swift/IMQuickSearch.swift
//
//  IMQuickSearch.swift
//  IMQuickSearch
//
//  Created by Ben Gordon on 11/25/14.
//  Copyright (c) 2014 Intermark. All rights reserved.
//

import UIKit

typealias IMQuickSearchFilterCompletion = (objects: [AnyObject]?) -> ()

class IMQuickSearch: NSObject {
    // MARK: - Properties
    var masterArray: [IMQuickSearchFilter] = []
    
    
    // MARK: - Init
    init(filters: [IMQuickSearchFilter]) {
        super.init()
        masterArray = filters
    }
    
    
    // MARK: - Filtered Objects
    func filteredObjects(value: String?) -> [AnyObject] {
        // Set Up
        var filteredSet = NSMutableSet()
        let filters = masterArray
        
        // Filter each object based on value
        for filter in filters {
            filteredSet.unionSet(filter.filteredObjects(value))
        }
        
        // Return it
        return filteredSet.allObjects
    }
    
    func asynchronouslyFilteredObjects(value: String?, completion: IMQuickSearchFilterCompletion) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), { () -> Void in
            let filteredObj = self.filteredObjects(value)
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                completion(objects: filteredObj)
            })
        })
    }
    
    
    // MARK: - Filter Management
    func add(filter: IMQuickSearchFilter) {
        masterArray.append(filter)
    }
    
    func remove(filter: IMQuickSearchFilter) {
        var filterToDeleteIndex = -1
        for (var i = 0; i < masterArray.count; i++) {
            let f = masterArray[i]
            if (f == filter) {
                filterToDeleteIndex = i
            }
        }
        
        if (filterToDeleteIndex >= 0) {
            masterArray.removeAtIndex(filterToDeleteIndex)
        }
    }
}

class IMQuickSearchFilter: NSObject {
    // MARK: - Properties
    var searchSet: NSSet = NSSet()
    var lastSearchSet: NSSet? = nil
    var lastSearchValue: NSString? = nil
    var keys: [String]? = nil
    
    
    // MARK: - Init
    init(searchArray: [AnyObject]!, keys: [String]) {
        super.init()
        self.searchSet = NSSet(array: searchArray)
        self.keys = keys
    }
    
    
    // MARK: - Filtering Objects
    func filteredObjects(value: String?) -> NSSet {
        if let v = value {
            // If value's length == 0, return all results
            if (NSString(string: v).length == 0) {
                return searchSet
            }
            
            // Set Up
            let shouldUseLastSearch = valueContains(v, containsValue: lastSearchValue)
            var newSearchSet = (lastSearchSet != nil && shouldUseLastSearch) ? lastSearchSet : searchSet
            
            // Create predicate
            var predicate = quickSearchPredicate(keys, value: v)
            var filteredSet = newSearchSet!.filteredSetUsingPredicate(predicate)
            
            // Save
            lastSearchSet = filteredSet
            lastSearchValue = v
        }
        
        // No value, return all sets
        return searchSet
    }
    
    
    // MARK: - Create predicate
    func quickSearchPredicate(keys: [String]?, value: String) -> NSPredicate {
        if let k = keys {
            var predicates: [NSPredicate] = []
            for key in k {
                var existsPredicate = NSPredicate({ (evalObject, bindings) -> Bool in
                    return evalObject.respondsToSelector(Selector(value))
                })
                var containsPredicate = NSPredicate(format: "(%K.description CONTAINS[cd] %@)", key, value)
                predicates.append(NSCompoundPredicate.andPredicateWithSubpredicates([existsPredicate, containsPredicate!]))
            }
            
            return NSCompoundPredicate.orPredicateWithSubpredicates(predicates)
        }
        
        return NSPredicate()
    }
    
    
    // MARK: - Check string
    func valueContains(value: String?, containsValue value2: String?) -> Bool {
        if (value == nil || value2 == nil) {
            return false
        }
        
        // Unwrap and test for 0 length
        let v1 = value!
        let v2 = value2!
        if (NSString(string: v2).length == 0) {
            return true
        }
        
        // Return predicate evaluation
        let p = NSPredicate(format: "SELF CONTAINS[cd] %@", v2)
        return p!.evaluateWithObject(v1)
    }
}